File: 692-top-k-frequent-words/top-k-frequent-words.cpp
class Solution {
public:
    vector<string> topKFrequent(vector<string>& words, int k) {
        unordered_map<string, int> mp;
        for (auto& word : words)
            mp[word]++;

        vector<vector<string>> bucket(words.size() + 1);
        for (auto& i : mp)
            bucket[i.second].push_back(i.first);

        vector<string> ans;
        for (int i = bucket.size() - 1; i >= 0; i--) {
            if (bucket[i].size() > 1) {
                sort(bucket[i].begin(), bucket[i].end());
            }
            for (int j = 0; j < bucket[i].size(); j++) {
                ans.push_back(bucket[i][j]);
                if (ans.size() == k)
                    return ans;
            }
        }
        return ans;
    }
};

File: 692-top-k-frequent-words/top-k-frequent-words.java
class Solution {
    public List<String> topKFrequent(String[] words, int k) {
        Map<String, Integer> map = new HashMap<>();
        for (String word : words) {
            map.put(word, map.getOrDefault(word, 0) + 1);
        }
        int maxi = 0;
        for (Integer val : map.values())
            maxi = Math.max(maxi, val);

        List<List<String>> bucket = new ArrayList<>();
        for (int i = 0; i <= maxi; i++)
            bucket.add(new ArrayList<>());

        for (String key : map.keySet()) {
            int freq = map.get(key);
            bucket.get(freq).add(key);
        }

        for (List<String> list : bucket) {
            Collections.sort(list, (a, b) -> a.compareTo(b));
        }

        List<String> res = new ArrayList<>();
        for (int i = maxi; i >= 0; i--) {
            for (String key : bucket.get(i)) {
                res.add(key);
                k--;
                if (k == 0)
                    break;
            }
            if (k == 0)
                break;
        }
        return res;
    }
}